spec/lib/cheers_leader_spec.rb

```diff
@@ -1,48 +1,118 @@
 require 'cheers_leader'
 
 describe CheerLeaderList do
-  before do
-    arr = [[2, "luan", 2, 2], [1, "tuan", 1, 1], [3, "mai", 2, 3]]  
+  before do 
+    leader_template = Struct.new(:rank, :name, :sent, :received)
+    arr = []
+    arr << leader_template.new(2, "tuan", 2, 2)
+    arr << leader_template.new(1, 'luan', 1, 1)
+    arr << leader_template.new(3, 'ngoc', 3, 3)
     @list = CheerLeaderList.new(arr)
   end
-  
-  context "Cheers Leader list informations" do 
-    it "By default, the Cheers Leader list is empty" do
-      expect(CheerLeaderList.new(nil).sort_by('')).to eq([])
+
+  context 'Cheers Leader list contain rank, name, sent and received information' do
+
+    it 'Return by default sorted by rank asc'
+    # DONT KNOW HOW IT NOT WORKING ?? 
+
+    it 'Container all information' do
+      sample = @list.list[0]
+      expect(sample.rank).to eq 2
+      expect(sample.name).to eq "tuan"
+      expect(sample.sent).to eq 2
+      expect(sample.received).to eq 2
     end
+  end
 
-    it "Cheers Leader list contain rank, name, sent and receiveed " do
-      sub_arr = @list.sort_by('')[0]
-      expect(sub_arr).to eq [1, "tuan", 1, 1]
+  context 'Sort feature for Cheers Leader list' do
+
+    it 'Can be sorted by rank asc' do 
+      result = []
+      # WHY these below NOT WORKING ??? 
+      # ranks = @list.sort_by_rank.each_with_object([]) {|x, ranks| ranks[x] = x.rank}
+      # result = @list.sort_by_rank.each {|x| result[x] << x.rank}
+      @list.sort_by_rank.each do |x|
+        result << x.rank
+      end
+      expect(result).to eq [1,2,3]
+    end
+    
+    it 'Can be sorted by rank desc' do 
+      result = []
+      @list.sort_by_rank(false).each do |x|
+        result << x.rank
+      end
+      expect(result).to eq [3,2,1]
     end
 
-  end
+    it 'Can be sorted by name asc' do 
+      result = []
+      @list.sort_by_name.each do |x|
+        result << x.name
+      end
+      expect(result).to eq ["luan", "ngoc", "tuan"]
+    end
+    
+    it 'Can be sorted by name desc' do 
+      result = []
+      @list.sort_by_name(false).each do |x|
+        result << x.name
+      end
+      expect(result).to eq ["tuan", "ngoc", "luan"]
```

Comment:
why not have a method called `@list.sorted_names` so you can focus on testing the behavior -- avoid code logic in test

Reply:
Thanks, will refactor later for this.